### ISSUE TITLE
Macro links do not function in shadow DOMs.

### ISSUE THREAD

**Issue by kwvanderlinde** [RPTools/maptool#2362]
**Describe the bug**
Clicking a macro link inside a shadow DOM does not cause the macro to run.

**To Reproduce**
1. Create a macro on a `lib:` token that just outputs a string. Label it `test`.
    ```
    Hello, world!
    ```
2. Create another macro to open an HTML5 frame.
3. Inside the frame place a macro link in the body.
4. Define a custom element that places its content in a shadow DOM, and add a macro link to it.
5. Example code:
    ```
    [frame5('Test Frame'): {
      <!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="utf-8">
          <script type="text/javascript">
          [r: "
            customElements.define('my-element', class extends HTMLElement {
              constructor() {
                super();
                this._shadowRoot = this.attachShadow({
                  mode: 'closed'   // or 'open', it doesn't matter
                });
              }

              connectedCallback() {
                let template = document.getElementById('my-element-template');
                const content = template.content.cloneNode(true);

                this._shadowRoot.appendChild(content);
              }
            });
          "]
          </script>
        </head>
        <body>
          [r: macroLink('Roll Me!', 'test@this', 'all')]

          <template id="my-element-template">
            [r: macroLink('Roll Me Too!', 'test@this', 'all')]
          </template>
          <my-element></my-element>
        </body>
      </html>
    }]
    ```
6. Click on the "Roll Me!" link and observe that "Hello, world!" is output to chat.
7. Click on the "Roll Me Too!" link and witness nothing happening.

**Expected behavior**
Clicking macro links inside a shadow DOM works just like clicking macro links outside of a shadow DOM.

**MapTool Info**
- Version: 1.8.0-beta.2
- Install: New

**Desktop (please complete the following information):**
 - OS: Linux Mint
 - Version 20

**Additional context**
I am working around this by duplicating the `MutationObserver` that MapTool uses to handle macro links that are not in shadow DOMs. If we add this code just before the `appendChild()` call, we get a functioning macro link:
```
const maptoolObserver = new MutationObserver(function(mutations, observer) {
	for(let mutation of mutations) {
		if (mutation.type === 'childList') {
			for (let i = 0; i < mutation.addedNodes.length; i++) {
				MapTool.handleAddedNode(mutation.addedNodes[i]);
			}
		}
	}
});
maptoolObserver.observe(this._shadowRoot, {
	attributes: false,
	characterData: false,
	childList: true,
	subtree: true
});
```

Also, while my example code uses `'closed'` shadow root, but the same behaviour occurs for `'open'` as well.

**Comment by Azhrei** [comment]
Since MapTool simply packages up WebKit as provided by the JavaFX WebView/WebEngine API, MT only supports what that API supports.  Since [this page](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) indicates such functionality is not available in Safari (and since Safari is based on WebKit), I'm going to guess that this is something MT cannot support until such time as the underlying WebKit layer is updated.

@Phergus I'm inclined to tag this **wontfix**, but you're managing the repo so I'll leave it to you to decide.

**Comment by Azhrei** [comment]
Perhaps my reply was premature.  I find that the live code examples on the previously linked page do work in a recent version of Safari, but they (1) don't require event callbacks and (2) are running in a more recent version of WebKit than MapTool uses.  Evaluating whether the issue is WebKit or a lack of support in MapTool is going to take more experimentation, and I'm not an expert on shadow DOM operations nor on WebKit plumbing in MapTool.

**Comment by kwvanderlinde** [comment]
I think the issue is that the MutationObserver that is used is only attached to the light DOM (I believe this is expected behaviour for a MutationObserver). Somehow it would need to observe each shadow DOM that gets added, like I did in my workaround. I think this *could* be done automatically for open shadow roots but I doubt it's possible for closed ones, as restricting script access is kind of the point of those.

I believe the only way to do it generally would be for WebView to allow custom handlers for each protocol. Which it sort of does (URLStreamHandlerFactory) but I don't think that permits the use case of a link that really acts more like a button.

Since this is a pretty niche, and there are workarounds (e.g., copying the MutationObserver, putting the links into light-side and slotting them in, otherwise scripting the links myself, etc.) I wouldn't be at all upset if this were to be marked "wontfix".

**Comment by cwisniew** [comment]
Unlikely to be fixed, see #3419 for more context
